Provider/src/main/java/com/spotify/confidence/AndroidLifecycleEventProducer.kt

```diff
@@ -0,0 +1,235 @@
+                }
+                put("url", ConfidenceValue.String(uri.toString()))
+            }
+            put("referrer", ConfidenceValue.String(getReferrer(activity).toString()))
+        }
+        coroutineScope.launch { eventsFlow.emit(Event("Deeplink", map)) }
+    }
+
+    override fun onActivityStopped(p0: Activity) {
+        p0.trackActivity("activity-stopped")
+    }
+
+    override fun onActivitySaveInstanceState(p0: Activity, p1: Bundle) {}
+
+    override fun onActivityDestroyed(p0: Activity) {
+        p0.trackActivity("activity-destroyed")
+    }
+
+    private fun trackApplicationLifecycleEvents() {
+        // Get the current version.
+        val packageInfo = packageInfo
+        val currentVersion = ConfidenceValue.String(packageInfo.versionName)
+        val currentBuild = ConfidenceValue.String(packageInfo.getVersionCode().toString())
+
+        // Get the previous recorded version.
+        val previousVersion = sharedPreferences
+            .getString(APP_VERSION, null)
+            ?.let(ConfidenceValue::String)
+
+        val previousBuild: ConfidenceValue.String? = sharedPreferences
+            .getString(APP_BUILD, null)
+            ?.let(ConfidenceValue::String)
+
+        val legacyPreviousBuild = sharedPreferences.getString(LEGACY_APP_BUILD, null)
+
+        // Check and track Application Installed or Application Updated.
+        if (previousBuild == null && legacyPreviousBuild == null) {
+            val message = mapOf("version" to currentVersion, "build" to currentBuild)
+            coroutineScope.launch { eventsFlow.emit(Event(APP_INSTALLED_EVENT, message)) }
+        } else if (currentBuild != previousBuild) {
+            val message = mapOf(
+                "version" to currentVersion,
+                "build" to currentBuild,
+                "previous_version" to (previousVersion ?: ConfidenceValue.String("")),
+                "previous_build" to (previousBuild ?: ConfidenceValue.String(""))
+            )
+            coroutineScope.launch { eventsFlow.emit(Event(APP_UPDATED_EVENT, message)) }
+        }
+
+        coroutineScope.launch {
+            sharedPreferences.edit().putString(APP_VERSION, currentVersion.string).apply()
+            sharedPreferences.edit().putString(APP_BUILD, currentBuild.string).apply()
+        }
+
+        coroutineScope.launch {
+            val message = mapOf("version" to currentVersion, "build" to currentBuild)
+            eventsFlow.emit(Event(APP_CREATED, message))
+        }
+    }
+
+    override fun events(): Flow<Event> = eventsFlow
+
+    override fun contextChanges(): Flow<Map<String, ConfidenceValue>> = contextFlow
+    override fun stop() {
+        application.unregisterActivityLifecycleCallbacks(this)
+        coroutineScope.launch(Dispatchers.Main) {
+            lifecycle.removeObserver(this@AndroidLifecycleEventProducer)
+        }
+    }
+
+    companion object {
+        private const val IS_FOREGROUND_KEY = "is_foreground"
```

Comment:
Note: This is a good example of a context that we probably don't want to re-resolve flags for, right?

Reply:
yes but after merging your changes we shouldn't be returning stale anymore and i think it's still nice to send these.